File: .tmp_py4013/batch_17_dump/3864_minimum_cost_to_partition_a_binary_string.js
// LeetCode 3864 - Minimum Cost To Partition A Binary String
// https://leetcode.com/problems/minimum-cost-to-partition-a-binary-string/

var minCost = function(s, encCost, flatCost) {
    const n = s.length;
    const pre = new Array(n + 1).fill(0);
    for (let i = 1; i <= n; i++) pre[i] = pre[i - 1] + (s.charCodeAt(i - 1) - 48);
    const dfs = (l, r) => {
        const x = pre[r] - pre[l];
        let res = x !== 0 ? (r - l) * x * encCost : flatCost;
        if ((r - l) % 2 === 0) {
            const m = Math.floor((l + r) / 2);
            res = Math.min(res, dfs(l, m) + dfs(m, r));
        }
        return res;
    };
    return dfs(0, n);
};

File: .tmp_rs4013/rust_check21/3864_minimum_cost_to_partition_a_binary_string.rs
struct Solution;
// LeetCode 3864 - Minimum Cost to Partition a Binary String
// https://leetcode.com/problems/minimum-cost-to-partition-a-binary-string/

impl Solution {
    pub fn min_cost(s: String, enc_cost: i32, flat_cost: i32) -> i64 {
        let n = s.len();
        let b = s.as_bytes();
        let mut pre = vec![0; n + 1];
        for i in 1..=n {
            pre[i] = pre[i - 1] + (b[i - 1] - b'0') as i32;
        }
        fn dfs(l: usize, r: usize, pre: &[i32], enc_cost: i32, flat_cost: i32) -> i64 {
            let x = pre[r] - pre[l];
            let mut res = if x != 0 {
                (r - l) as i64 * x as i64 * enc_cost as i64
            } else {
                flat_cost as i64
            };
            if (r - l) % 2 == 0 {
                let m = (l + r) / 2;
                res = res.min(dfs(l, m, pre, enc_cost, flat_cost) + dfs(m, r, pre, enc_cost, flat_cost));
            }
            res
        }
        dfs(0, n, &pre, enc_cost, flat_cost)
    }
}
